cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/ETLSparkProgram.java

```diff
@@ -265,5 +330,61 @@ public MapFunction(SparkContext sparkContext, @Nullable String aggregatorName) {
 
       return transformExecutorFactory.create(pipelinePhase);
     }
+
+    @Override
+    protected KeyValue<Object, T> computeInputForExecutor(Tuple2<Object, T> input) {
+      return new KeyValue<>(input._1(), input._2());
+    }
+  }
+
+  /**
+   * Used for the transform after a SparkTransform. Otherwise, MapFunction only operates on RDD of JavaPairRDD.
+   * In other words, it does not handle translation from Tuple to KeyValue, but directly sends the RDD type
+   * to the TransformExecutor.
+   * This allows operations on JavaRDD of single type. Handles no aggregation functionality, because it should not
+   * be used in a phase with aggregations.
+   */
+  public static final class SingleTypeRDDMapFunction
+    extends TransformExecutorFunction<Object, Object, String, Object> {
+
+    public SingleTypeRDDMapFunction(SparkContext sparkContext, @Nullable String pipelineStr) {
+      super(sparkContext, pipelineStr);
+    }
+
+    @Override
+    protected Iterable<Tuple2<String, Object>> getOutput(TransformResponse transformResponse) {
+      return mapFunctionGetOutput(transformResponse);
+    }
+
+    @Override
+    protected TransformExecutor<Object> initialize(BatchPhaseSpec phaseSpec,
+                                                   PipelinePluginInstantiator pluginInstantiator) throws Exception {
+      TransformExecutorFactory<Object> transformExecutorFactory =
+        new SparkTransformExecutorFactory<>(pluginContext, pluginInstantiator, metrics,
+                                            logicalStartTime, runtimeArgs, false);
+
+      PipelinePhase pipelinePhase = phaseSpec.getPhase();
+      return transformExecutorFactory.create(pipelinePhase);
+    }
+
+    @Override
+    protected Object computeInputForExecutor(Object input) {
+      return input;
+    }
   }
+
+  // common getOutput implementation between MapFunction and SingleTypeRDDMapFunction
+  // seems unnecessarily complex to have an additional parent class to put between those two classes just for this
+  // implementation
+  private static Iterable<Tuple2<String, Object>> mapFunctionGetOutput(TransformResponse transformResponse) {
```

Comment:
you could make SingleTypeRDDMapFunction generic and have MapFunction extend it with KeyValue type and overriden computeInputForExecutor